code/modules/hydroponics/biogenerator.dm

```diff
@@ -1,337 +1,354 @@
+		"has_plants" = length(stored_plants) ? TRUE : FALSE
+	)
+	if(container)
+		data["container"] = TRUE
+		data["container_curr_reagents"] = container.reagents.total_volume
+		data["container_max_reagents"] = container.reagents.maximum_volume
+	else
+		data["container"] = FALSE
+
+	return data
 
-/obj/machinery/biogenerator/proc/activate()
-	if(usr.stat != 0)
+/obj/machinery/biogenerator/ui_static_data(mob/user)
+	return list("product_list" = product_list)
+
+/obj/machinery/biogenerator/ui_act(action, list/params)
+	if(..())
 		return
-	if(stat != 0) //NOPOWER etc
+
+	. = TRUE
+	switch(action)
+		if("activate")
+			activate(usr)
+		if("detach_container")
+			detach_container()
+		if("eject_plants")
+			eject_plants()
+		if("create")
+			var/datum/design/D = files.known_designs[params["id"]]
+			if(!D)
+				return
+			var/amount = clamp(text2num(params["amount"]), 1, 10)
+			create_product(D, amount)
+
+/**
+ * Tells the biogenerator to grind up any stored plants, and produce an appropriate amount of biomass.
+ *
+ * Argumens:
+ * * mob/user - the mob who activated the biogenerator
+ */
+/obj/machinery/biogenerator/proc/activate(mob/user)
+	if(stat & (NOPOWER | BROKEN))
 		return
 	if(processing)
-		to_chat(usr, "<span class='warning'>The biogenerator is in the process of working.</span>")
+		to_chat(user, "<span class='warning'>[src] is currently processing!</span>")
 		return
+
+	processing = TRUE
+	SStgui.update_uis(src)
+	update_icon()
+
 	var/S = 0
-	for(var/obj/item/reagent_containers/food/snacks/grown/I in contents)
+	for(var/obj/item/reagent_containers/food/snacks/grown/plant as anything in stored_plants)
 		S += 5
-		if(I.reagents.get_reagent_amount("nutriment")+I.reagents.get_reagent_amount("plantmatter") < 0.1)
-			points += 1*productivity
+		if(plant.reagents.get_reagent_amount("nutriment") + plant.reagents.get_reagent_amount("plantmatter") < 0.1)
+			biomass += productivity
 		else
-			points += (I.reagents.get_reagent_amount("nutriment")+I.reagents.get_reagent_amount("plantmatter"))*10*productivity
-		qdel(I)
-	if(S)
-		processing = 1
-		update_icon()
-		updateUsrDialog()
-		playsound(loc, 'sound/machines/blender.ogg', 50, 1)
-		use_power(S*30)
-		sleep(S+15/productivity)
-		processing = 0
-		update_icon()
-	else
-		menustat = "void"
-	return
-
-/obj/machinery/biogenerator/proc/check_cost(list/materials, multiplier = 1, remove_points = 1)
-	if(materials.len != 1 || materials[1] != MAT_BIOMASS)
-		return 0
-	if (materials[MAT_BIOMASS]*multiplier/efficiency > points)
-		menustat = "nopoints"
-		return 0
-	else
-		if(remove_points)
-			points -= materials[MAT_BIOMASS]*multiplier/efficiency
-		update_icon()
-		updateUsrDialog()
-		return 1
+			biomass += (plant.reagents.get_reagent_amount("nutriment") + plant.reagents.get_reagent_amount("plantmatter")) * 10 * productivity
+		qdel(plant)
+
+	playsound(loc, 'sound/machines/blender.ogg', 50, 1)
+	use_power(S * 30)
+	sleep(S + 15 / productivity)
+	stored_plants = list()
```

Comment:
make the list clearance happen before `sleep()`, not after (it only has qdel'ed objects right after the loop; no need to hold on to them across a sleep boundary and risk harddels)
(not a suggestion, because github hates me)

Also it's probably better to do `stored_plants.Cut()` instead of `=list()` - that does not create a new object, so is imperceptably faster